lib/recursive-methods.rb

```diff
@@ -1,49 +1,137 @@
+  end 
+  
+  return arr.join('')
+end
+
+# Time complexity: O(n), with n being the length of the string parameter
+# Space complexity: O(n), because of the recursive stack; size of the stack depends on the length of the string.
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_helper(s, 0 , s.length - 1)  
+end
+
+def reverse_helper(s, first, last)
+  if first < last
+    temp = s[first]
+    s[first] = s[last]
+    s[last] = temp
+
+    return reverse_helper(s, first + 1, last - 1)
+  end
+
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) with n being how high the value of the integer is
+# Space complexity: O(n), because of the recursive stack; size of stack depends on value of parameter
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+
+  return (2 + (2 * (n - 1)))
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), with n being the length of the string parameter
+# Space complexity: O(n), because of the recursive stack; size of the stack depends on the length of the string.
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length == 0
+  return false if s.length % 2 != 0
+
+  return nested_helper(s)
+end
+
+def nested_helper(s, i = 0, j = (s.length - 1))
+  if i < j
+    if (s[i] != '(' || s[j] != ')')
+      return false
+    end
+
+    return nested_helper(s, i + 1, j - 1)
+  end
+
+  return true
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), with n being the size of the array.
+# Space complexity: O(n^2), because of the recursive stack involving an array.
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.length == 0
+
+  return search_helper(array, value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def search_helper(a, v, i = 0)
+  if i < a.length
+    if a[i] == v
+      return true
+    end
+
+    return search_helper(a, v, i + 1)
+  end
+
+  return false
+end
+
+# Time complexity: O(n), with n being the length of the string parameter
+# Space complexity: O(n), because of the recursive stack; size of the stack depends on the length of the string.
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length <= 1
+
+  return pal_helper(s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def pal_helper(s, i = 0, j = (s.length - 1))
+  if i < j
+    return false if (s[i] != s[j])
+    return pal_helper(s, i + 1, j - 1)
+  end
+
+  return true
+end
+
+# Time complexity: O(n), with n being the length in digits of the shortest number.
+# Space complexity: O(n), because of the recursive stack; size of the stack depends number of digit in the shortest num.
 def digit_match(n, m)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if (n % 10 != m % 10)
```

Comment:
So 1337 and 1338 will have a digit match of 0, because the ones digit doesn't match?